### PR TITLE
Update commit filtering and diff view for file-level permissions to better reflect P4 behavior

### PR DESCRIPTION
**Pull request by mollylogue**
This PR makes two main changes:
1) Updates sub-repo filtering when listing commits to return a commit which modifies _any_ file the user has access to (previously the user had have access to _all_ files modified in the commit to view the commit)
2) Updates the handling of git diffs to ensure that only the files which the user has access to in the commit/range of commits are shown.

Outstanding questions:
- Do we need to do filtering for [DiffSymbols](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/gitserver/commands.go?L331:29#tab=references) ?

## Test plan
- Existing/updated/added unit tests all pass
- Local testing w/ github.com/mollylogue/testing repo using the `path_excludes` rule `**_test.go` demonstrates expected behavior for listing commits and viewing diffs

### CODE REVIEW

**Comment by ryanslade** [comment]
> Do we need to do filtering for [DiffSymbols](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/gitserver/commands.go?L331:29#tab=references) ?

I think we can skip that for now since it's consumed by the symbols service, not a user and also it only exposes file names not contents.

File: internal/gitserver/commands.go [hunk 1]
```diff
@@ -91,14 +91,31 @@ func (c *ClientImplementor) Diff(ctx context.Context, opts DiffOptions) (*DiffFi
...
+}
+
+type diffFileIteratorFilter func(fileName string) (bool, error)
+
+func getFilterFunc(ctx context.Context, checker authz.SubRepoPermissionChecker, repo api.RepoName) diffFileIteratorFilter {
+	if !authz.SubRepoEnabled(checker) {
+		return nil
+	}
+	return func(fileName string) (bool, error) {
+		shouldFilterNew, err := authz.FilterActorPath(ctx, checker, actor.FromContext(ctx), repo, fileName)
```
Comment (ryanslade) [inline internal/gitserver/commands.go:113]: nit: Why is it called `shouldFilterNew` and not `shouldFilter`?

File: internal/gitserver/commands_test.go [hunk 2]
```diff
@@ -1621,16 +1728,24 @@ func TestCommits_SubRepoPerms(t *testing.T) {
...
 
 	tests := map[string]struct {
 		wantCommits   []*gitdomain.Commit
 		opt           CommitsOptions
 		wantTotal     uint
 		noAccessPaths []string
 	}{
-		"if no read perms on file should filter out commit": {
-			wantTotal: 1,
+		"if no read perms on at least on file in the commit should filter out commit": {
```
Comment (ryanslade) [inline internal/gitserver/commands_test.go:1739]:
```suggestion
		"if no read perms on at least one file in the commit should filter out commit": {
```

**Review by ryanslade — APPROVED** [review]
Awesome!

One thing we should also look into is what happens with diff searches

File: internal/gitserver/commands.go (hunk 1 above)
Comment (mollylogue) [reply internal/gitserver/commands.go:113]: Ah. This is a holdover from when I was checking both the new and old filepaths. Just forgot to rename.

**Comment by mollylogue** [comment]
> This PR makes two main changes:
> 
> 1. Updates sub-repo filtering when listing commits to return a commit which modifies _any_ file the user has access to (previously the user had have access to _all_ files modified in the commit to view the commit)
> 2. Updates the handling of git diffs to ensure that only the files which the user has access to in the commit/range of commits are shown.
> 
> ## Test plan
> * Existing/updated/added unit tests all pass
> * Local testing w/ github.com/mollylogue/testing repo using the `path_excludes` rule `**_test.go` demonstrates expected behavior for listing commits and viewing diffs



> Awesome!
> 
> One thing we should also look into is what happens with diff searches

Ah good point. I will look into that/potentially address in a separate PR.

**Pull request merged**